website_cms/models/cms_media.py

```diff
@@ -0,0 +1,313 @@
+    def create(self, vals):
+        """Override to keep link w/ page resource and handle url.
+
+        `ir.attachment` have a weak relation w/ related resources.
+        We want the user to be able to select a page,
+        but on the same time we need to keep this in sync w/
+        attachment machinery. There you go!
+        """
+        if vals.get('page_id') is not None:
+            vals['res_id'] = vals.get('page_id')
+            vals['res_model'] = vals.get('page_id') and 'cms.page'
+        # TODO: use noembed service
+        url = vals.get('url')
+        if url and 'youtube' in url:
+            vals['url'] = to_embed_url(url)
+        return super(CMSMedia, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        """Override to keep link w/ page resource."""
+        if vals.get('page_id') is not None:
+            vals['res_id'] = vals.get('page_id')
+            vals['res_model'] = vals.get('page_id') and 'cms.page'
+        # TODO: use noembed service
+        url = vals.get('url')
+        if url and 'youtube' in url:
+            vals['url'] = to_embed_url(url)
+        return super(CMSMedia, self).write(vals)
+
+    @api.model
+    def is_image(self, mimetype=None):
+        """Whether this is an image."""
+        mimetype = mimetype or self.mimetype
+        if mimetype:
+            return mimetype in IMAGE_TYPES
+        # mimetype is computed on create
+        if self.url:
+            return guess_mimetype(self.url) in IMAGE_TYPES
+
+    @api.model
+    def is_video(self):
+        """Whether this is a video."""
+        return self.mimetype in VIDEO_TYPES
+
+    # TODO: we should use services like http://noembed.com
+    @api.multi
+    @api.onchange('url', 'datas')
+    def _preload_preview_image(self):
+        """Preload preview image based on url and mimetype."""
+        for item in self:
+            # use _compute_mimetype from ir.attachment
+            # that unfortunately is used only on create
+            values = {}
+            for fname in ('datas_fname', 'url', 'datas'):
+                values[fname] = getattr(item, fname)
+            item.mimetype = self._compute_mimetype(values)
+
+            if item.is_image():
+                if item.url:
+                    image_content = download_image_from_url(item.url)
+                else:
+                    image_content = item.datas
+                item.image = image_content
+            if item.url and 'youtube' in item.url:
+                image_content = download_image_from_url(
+                    to_preview_url(item.url))
+                item.image = image_content
+
+
+class CMSMediaCategory(models.Model):
+    """Model of a CMS media category."""
+
+    _name = 'cms.media.category'
+    _inherit = 'website.orderable.mixin'
+    _description = 'CMS Media Category'
+
+    name = fields.Char(
+        'Name',
+        required=True,
+        translate=True,
+    )
+    mimetypes = fields.Text(
+        'Mimetypes',
```

Comment:
because I didn't want to add an extra model just for that, and adding new mimetypes is faster in this way. Which are the real benefits here?